keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -907,22 +917,26 @@ Problem examples:
 | CRD is in multiple clusters, but schema differs        | If policy author is aware of the schema variations, can they write policies that work for all the variations? |
 | Validation of an aggregated API server type            | Main API server does not have type definitions |
 
-Due to these complications, we have decided to evalute CEL expressions
-dynamically. Informational type checking will be provided (except for aggregated
-API server types), but will be surfaced only as warnings. See "Alternatives
-Considered" section for details of all the alternatives we reviewed when
-selecting this approach.
+Due to these complications, Type checking will be optional which can be activated by an expression-level
+`TypeCheckingMode` field. Possible modes are:
 
-Type checking is still performed for all expressions where a GVK can be matched
-to type check against, resulting in warnings, e.g.:
+- `Never` Do not type check the referred object or the param any time. 
+  Treat `object`, `oldObject`, and `params` as dynamic-typed unstructured objects. 
+  This behavior matches that of Kubernetes 1.26
+- `Evaluation` Type check the referred `object`, `oldObject`, and `params` when the expression is being compiled and evaluated after the policy is created. 
+  If the type check fails, mark the policy as misconfigured. Do not type check when the policy is being created or updated.
```

Comment:
The best we can do is to return the type check errors as part of the failure message, taking the place of per-expression failure message. For example
```
- expression: "object.spec.replicas > 1" 
  message: "must be replicated"
```
This will compile and eval the expression, and return "must be replicated", 
```
- expression: "object.replicas > 1" 
  message: "must be replicated"
```
This will compile-error, and return the error from the compiler instead of the message.

Reply:
For the second point. `Never` should mean never, this is the explicit opt-out. Would you think "errors are reported but don't have any runtime implications" could be a new mode? Like the real "Informative" mode. We can make this default instead of `Never`.

However, making it informative comes with the previously mentioned how-to-report problem. I am thinking of extending the status beyond `misconfigured` or not, for example, by adding a specific field for type checking results.

Reply:
> Would you think "errors are reported but don't have any runtime implications" could be a new mode? Like the real "Informative" mode. We can make this default instead of `Never`.

Yeah, an "Informative" mode could be useful for policy authors that want a simple way to check for type errors (it should be possible to provide type errors in the status of a policy definition even if there are no policy bindings defined). It could also be useful for cluster administrators to check for compatibility of a policy with a cluster before creating any bindings.  I would be okay with this being the default instead of `Never`.

> However, making it informative comes with the previously mentioned how-to-report problem. I am thinking of extending the status beyond `misconfigured` or not, for example, by adding a specific field for type checking results.

This sounds reasonable to me.